{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "This file adds the columns 'cluster_hap' and 'cluster_hap_imputed' to a data table. The column 'cluster_hap' contains parent of origin assignments for each read based on the chromosome territory it belongs to, without using information from other territories. The column 'cluster_hap_imputd' contains parent of origin assignments that leverage the parent of origin information about homologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adjust datapath to reflect the location of data in your file system\n",
    "datapath = \"/Users/paulreginato/Google Drive/genome seq in situ/Analysis/191202_embryo_analysis/embryo_seq_data_tables/201125_embryo_data_table_Y_nonchromosomal_fixed2.csv\"\n",
    "source_data = pd.read_csv(datapath,index_col=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Call parent of origin for each cluster based on the SNP assignments for reads in the cluster by tallying\n",
    "# the number of SNPs from each parent in the cluster and making the call based on majority rules. If the cluster\n",
    "# cannot be assigned to a parent because it does not contain SNPs or contains the same number of SNPs from each \n",
    "# parent, leave it unclassified. If the cluster is an X chromosome and there is only one cluster in the cell,\n",
    "# call parent of origin as maternal. Add the column 'cluster_hap' to indicate the parent call, which will be\n",
    "# 0 if maternal, 1 if paternal, and -1 if unclassified. Add the column 'single_snp' to indicate whether the \n",
    "# cluster had only one SNP in the cluster. Add columns 'num_hap0' and 'num_hap1', which will keep track of the\n",
    "# number of SNPs from each parent in the cluster. Takes a data table\n",
    "def call_raw_haps(data):\n",
    "    \n",
    "    new_data=data.copy()\n",
    "    \n",
    "    #create the new columns\n",
    "    new_data['cluster_hap']=-1\n",
    "    new_data['single_snp']=0\n",
    "    new_data['num_hap0']=0\n",
    "    new_data['num_hap1']=0\n",
    "\n",
    "    \n",
    "    #loop through all clusters in all cells\n",
    "    cell_indices = new_data.cell_index.unique()\n",
    "    \n",
    "    for cell_index in cell_indices:\n",
    "        cell = new_data.loc[new_data.cell_index==cell_index]\n",
    "        \n",
    "        chr_nums = cell.chr.unique()\n",
    "            \n",
    "        for chr_num in chr_nums:\n",
    "            chrom = cell.loc[cell.chr==chr_num]\n",
    "            clust_nums = chrom.cluster.unique()\n",
    "            \n",
    "            #If there is only one cluster and it is an X chromosome, assign 'cluster_hap' \n",
    "            # to 0 (maternal).\n",
    "            if ((len(clust_nums)==1)&(chr_num==20)):\n",
    "                new_data.loc[chrom.index,'cluster_hap']=0\n",
    "                continue\n",
    "            \n",
    "            for clust_num in clust_nums:  \n",
    "                clust = chrom.loc[chrom.cluster==clust_num]\n",
    "                \n",
    "                #get the number of SNPs for each hap in the cluster\n",
    "                num_hap0=len(clust.loc[clust.hap_assignment==0])\n",
    "                num_hap1=len(clust.loc[clust.hap_assignment==1])\n",
    "                \n",
    "                #record the number of SNPs in each hap\n",
    "                new_data.loc[clust.index,'num_hap0']=num_hap0\n",
    "                new_data.loc[clust.index,'num_hap1']=num_hap1\n",
    "                \n",
    "                #compare the number of SNPs in each hap and assign haplotype based on majority\n",
    "                if num_hap0>num_hap1:\n",
    "                    new_data.loc[clust.index,'cluster_hap']=0\n",
    "                    win_num = num_hap0\n",
    "                    lose_num = num_hap1\n",
    "                elif num_hap1>num_hap0:\n",
    "                    new_data.loc[clust.index,'cluster_hap']=1\n",
    "                    win_num = num_hap1\n",
    "                    lose_num = num_hap0\n",
    "                    \n",
    "                #fill in the column 'single_snp'\n",
    "                if win_num==1:\n",
    "                    new_data.loc[clust.index,'single_snp']=1\n",
    "                    \n",
    "    return new_data\n",
    "\n",
    "\n",
    "# revise the raw haplotype calls for each chromosome based on the identities of the other clusters. Add\n",
    "# the new column 'cluster_hap_imputed'. Rules:\n",
    "#1. If the cluster number is -1, assign 'cluster_hap_imputed' to -1.\n",
    "#2. If one cluster is assigned to parent of origin and the other is not, call the unclassified cluster as the \n",
    "# opposite of the one that is classified.\n",
    "#3. If the parent of origin calls are the same but one of the clusters contained only one SNP and the other \n",
    "#had multiple, change the one with only one SNP.\n",
    "#4. If the parent of origin calls are the same and Rule 3 can't be applied, assign 'cluster_hap_imputed' to -1.\n",
    "#5. If the chromosome is aneuploid, assign 'cluster_hap_imputed' to the same value as the raw assignment.\n",
    "#6. If there is only one cluster and it is an X chromosome, assign 'cluster_hap_imputed' to 0 (maternal).\n",
    "\n",
    "\n",
    "#takes a dataframe outputted by call_raw_haps()\n",
    "\n",
    "def revise_raw_hap_calls(data):\n",
    "    num_disagreements = 0\n",
    "    new_data=data.copy()\n",
    "    new_data['cluster_hap_imputed']=-1\n",
    "    \n",
    "    #loop through every chromosome in every cell\n",
    "    cell_indices = new_data.cell_index.unique()\n",
    "    \n",
    "    for cell_index in cell_indices:\n",
    "        cell = new_data.loc[new_data.cell_index==cell_index]\n",
    "        \n",
    "        chr_nums = cell.chr.unique()\n",
    "            \n",
    "        for chr_num in chr_nums:\n",
    "            chrom = cell.loc[cell.chr==chr_num]\n",
    "            clust_nums = chrom.cluster.unique()\n",
    "            \n",
    "            #Rule 1: if any of the clusters have cluster number -1, assign 'cluster_hap_imputed' to -1 \n",
    "            if -1 in list(chrom.cluster):\n",
    "                continue\n",
    "            \n",
    "            #Rule 6: If there is only one cluster and it is an X chromosome, assign 'cluster_hap_imputed' \n",
    "            # to 0 (maternal).\n",
    "            if ((len(clust_nums)==1)&(chr_num==20)):\n",
    "                new_data.loc[chrom.index,'cluster_hap_imputed']=0\n",
    "                continue\n",
    "            \n",
    "            #Rule 5: if there are more than two clusters, assign 'cluster_hap_imputed' to the same value \n",
    "            #as the raw assignment.\n",
    "            if len(clust_nums)!=2:\n",
    "                for clust_num in clust_nums:\n",
    "                    clust = chrom.loc[chrom.cluster==clust_num]\n",
    "                    clust_call = clust.iloc[0].cluster_hap\n",
    "                    new_data.loc[clust.index,'cluster_hap_imputed']=clust_call\n",
    "                continue\n",
    "                    \n",
    "            #If there are two clusters and neither are assigned cluster number -1, apply the other rules\n",
    "            \n",
    "            #Find the parent of origin assignments for each cluster\n",
    "            clust1 = chrom.loc[chrom.cluster==1]\n",
    "            clust2 = chrom.loc[chrom.cluster==2]\n",
    "            c1_call = clust1.iloc[0].cluster_hap\n",
    "            c2_call = clust2.iloc[0].cluster_hap\n",
    "            \n",
    "            #Rule 2: if the raw parent of origin calls are different leave them the same, unless one is \n",
    "            # unclassified, in which case update it to match the one that is classified. \n",
    "            if c1_call!=c2_call:\n",
    "                if c1_call == -1:\n",
    "                    new_data.loc[clust2.index,'cluster_hap_imputed']=c2_call\n",
    "                    new_data.loc[clust1.index,'cluster_hap_imputed']=abs(c2_call-1)\n",
    "                elif c2_call == -1:\n",
    "                    new_data.loc[clust1.index,'cluster_hap_imputed']=c1_call\n",
    "                    new_data.loc[clust2.index,'cluster_hap_imputed']=abs(c1_call-1)\n",
    "                else:\n",
    "                    new_data.loc[clust1.index,'cluster_hap_imputed']=c1_call\n",
    "                    new_data.loc[clust2.index,'cluster_hap_imputed']=c2_call\n",
    "                    \n",
    "                continue\n",
    "            \n",
    "            #Rule 4: if the raw parent of origin calls are the same and both clusters had more than one SNP,\n",
    "            # assign 'cluster_hap_imputed' to -1\n",
    "            if ((clust1.iloc[0]['single_snp']==0)&(clust2.iloc[0]['single_snp']==0)):\n",
    "                continue\n",
    "            \n",
    "            #Rule 3: if the parent of origin calls are the same but one of the clusters contained only one SNP \n",
    "            #and the other had multiple, change the one with only one SNP.\n",
    "            if ((clust1.iloc[0]['single_snp']==1)&(clust2.iloc[0]['single_snp']!=1)):\n",
    "                new_data.loc[clust2.index,'cluster_hap_imputed']=c2_call\n",
    "                new_data.loc[clust1.index,'cluster_hap_imputed']=abs(c2_call-1)\n",
    "                continue\n",
    "            \n",
    "            if ((clust2.iloc[0]['single_snp']==1)&(clust1.iloc[0]['single_snp']!=1)):\n",
    "                new_data.loc[clust1.index,'cluster_hap_imputed']=c1_call\n",
    "                new_data.loc[clust2.index,'cluster_hap_imputed']=abs(c1_call-1)\n",
    "                continue\n",
    "                \n",
    "    return new_data\n",
    "\n",
    "#Annotates the parent of origin assignments in the full dataset. Assigns outliers to the parent of origin\n",
    "#of their associated cluster in the cluster_hap and cluster_hap_imputed columns. All read that are not part of a\n",
    "#chromosome cluster are assigned -1 in those columns. Takes an unfiltered data table ('full_data') and annotates it\n",
    "#with the parent of origin data from a filtered table, ie the output of revise_raw_hap_calls ('assigned_data').\n",
    "def assign_haps_to_outliers(full_data,assigned_data):\n",
    "    new_data = full_data.copy()\n",
    "    new_data['cluster_hap']=-1\n",
    "    new_data['cluster_hap_imputed']=-1\n",
    "    \n",
    "    #go through each cluster in each cell in the assigned_data. Map the parent of origin assignments from each\n",
    "    #cluster in assigned_data to the same cluster in full_data, including outliers. This allows outliers in the\n",
    "    #full_data to be assigned parent of origin, even though they were not used to decide the assignments.\n",
    "    \n",
    "    cell_indices = assigned_data.cell_index.unique()\n",
    "    for ci in cell_indices:\n",
    "        cell = assigned_data.loc[assigned_data.cell_index==ci]\n",
    "        chr_nums = cell.chr.unique()\n",
    "        for chr_num in chr_nums:\n",
    "            chrom = cell.loc[cell.chr==chr_num]\n",
    "            clust_nums = chrom.cluster.unique()\n",
    "            for clust_num in clust_nums:\n",
    "                clust = chrom.loc[chrom.cluster==clust_num]\n",
    "                clust_hap = clust.iloc[0].cluster_hap\n",
    "                clust_hap_imputed = clust.iloc[0].cluster_hap_imputed\n",
    "                \n",
    "                new_data_cluster = new_data.loc[(new_data.cell_index==ci)&(new_data.chr==chr_num)&\n",
    "                                               (new_data.cluster==clust_num)]\n",
    "                new_data.loc[new_data_cluster.index,'cluster_hap']=clust_hap\n",
    "                new_data.loc[new_data_cluster.index,'cluster_hap_imputed']=clust_hap_imputed\n",
    "    \n",
    "    return new_data\n",
    "\n",
    "\n",
    "#Populates manual parent of origin assignments for zygotes. Cluster assignment 1 corresponds to maternal origin,\n",
    "# and cluster assignmennt 2 corresponds to paternal origin.\n",
    "def assign_manual_zygote_calls(data):\n",
    "    new_data = data.copy()\n",
    "    zygotes = data.loc[data.stage=='zygote']\n",
    "    new_data.loc[zygotes.index,'cluster_hap']=zygotes.cluster-1\n",
    "    new_data.loc[zygotes.index,'cluster_hap_imputed']=zygotes.cluster-1\n",
    "    \n",
    "    return new_data\n",
    "\n",
    "#Returns a dataframe with only euploid chromosomes\n",
    "def get_euploid(data):\n",
    "    data_c=data.copy()\n",
    "    cell_indices=data.cell_index.unique()\n",
    "    for cell_index in cell_indices:\n",
    "        cell = data.loc[data.cell_index==cell_index]\n",
    "        chr_nums=cell.chr.unique()\n",
    "        for chr_num in chr_nums:\n",
    "            chrom=cell.loc[cell.chr==chr_num]\n",
    "            num_clusters = len(chrom.cluster.unique())\n",
    "            if ((chr_num==20)&(num_clusters==1)):\n",
    "                continue\n",
    "            if num_clusters!=2:\n",
    "                data_c=data_c.drop(index=chrom.index)\n",
    "    return data_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get chromosome data and remove outliers and chromosomes with overlapping clusters\n",
    "chromosomes = source_data.loc[(source_data.chr<21)&(source_data.inlier==1)&(source_data.cluster!=-1)]\n",
    "\n",
    "#remove aneuploid chromosomes\n",
    "euploid = get_euploid(chromosomes)\n",
    "\n",
    "#get raw parent of origin estimates for each cluster, without using information from other clusters. \n",
    "raw_haps = call_raw_haps(euploid)\n",
    "\n",
    "#Revise parent of origin assignments using info from other clusters. \n",
    "revised_haps = revise_raw_hap_calls(raw_haps)\n",
    "\n",
    "#Transfer the parent of origin calls to the full data, such that outliers also get assigned parent of origin\n",
    "#based on the cluster they are associated with.\n",
    "annotated_data = assign_haps_to_outliers(source_data,revised_haps)\n",
    "\n",
    "#Assign manual zygote parent of origin assignments to the cluster_hap and cluster_hap_imputed columns\n",
    "annotated_data = assign_manual_zygote_calls(annotated_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write new annotations to file\n",
    "annotated_data.to_csv(\"../data/Table_S1_pgp1_data_table.csv\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
